File: javascript/src/method/getOrNull.ts
/*******************************************************************************
 Copyright (c) 2023-2024. Jonathan Bédard ~ JóôòKiwi

 This project is free to use.
 All the right is reserved to the author of this project.
 ******************************************************************************/

import type {Nullable, NullOr} from "@joookiwi/type"

import type {CollectionHolder}           from "../CollectionHolder"
import type {MinimalistCollectionHolder} from "../MinimalistCollectionHolder"

import {isArray}                       from "./isArray"
import {isCollectionHolder}            from "./isCollectionHolder"
import {isArrayByStructure}            from "./isArrayByStructure"
import {isCollectionHolderByStructure} from "./isCollectionHolderByStructure"

//#region -------------------- Facade methods --------------------

/**
 * Get the element at the specified index in the {@link collection}
 * or <b>null</b> if it is over the {@link size}
 *
 * @param collection The {@link Nullable nullable} collection ({@link MinimalistCollectionHolder}, {@link CollectionHolder} or {@link ReadonlyArray Array})
 * @param index      The index to retrieve a value
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/get-or-null.html Kotlin getOrNull(index)
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/element-at-or-null.html Kotlin elementAtOrNull(index)
 * @canReceiveNegativeValue
 * @extensionFunction
 */
export function getOrNull<const T, >(collection: Nullable<| MinimalistCollectionHolder<T> | readonly T[]>, index: number,): NullOr<T> {
    if (collection == null)
        return null
    if (isCollectionHolder<T>(collection,))
        return getOrNullByCollectionHolder(collection, index,)
    if (isArray(collection,))
        return getOrNullByArray(collection, index,)
    if (isCollectionHolderByStructure<T>(collection,))
        return getOrNullByCollectionHolder(collection, index,)
    if (isArrayByStructure<T>(collection,))
        return getOrNullByArray(collection, index,)
    return getOrNullByMinimalistCollectionHolder(collection, index,)
}


/**
 * Get the element at the specified index in the {@link collection}
 * or <b>null</b> if it is over the {@link size}
 *
 * @param collection The {@link Nullable nullable} {@link MinimalistCollectionHolder collection}
 * @param index      The index to retrieve a value
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/get-or-null.html Kotlin getOrNull(index)
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/element-at-or-null.html Kotlin elementAtOrNull(index)
 * @canReceiveNegativeValue
 * @extensionFunction
 */
export function getOrNullByMinimalistCollectionHolder<const T, >(collection: Nullable<MinimalistCollectionHolder<T>>, index: number,): NullOr<T> {
    if (collection == null)
        return null
    if (Number.isNaN(index,))
        return null
    if (index == Number.NEGATIVE_INFINITY)
        return null
    if (index == Number.POSITIVE_INFINITY)
        return null

    const size = collection.size
    if (size == 0)
        return null
    if (index >= size)
        return null
    if (index >= 0)
        return collection.get(index,)

    const indexToRetrieve = index + size
    if (indexToRetrieve < 0)
        return null
    if (indexToRetrieve > size)
        return null
    return collection.get(indexToRetrieve,)

}

/**
 * Get the element at the specified index in the {@link collection}
 * or <b>null</b> if it is over the {@link size}
 *
 * @param collection The {@link Nullable nullable} {@link CollectionHolder collection}
 * @param index      The index to retrieve a value
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/get-or-null.html Kotlin getOrNull(index)
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/element-at-or-null.html Kotlin elementAtOrNull(index)
 * @canReceiveNegativeValue
 * @extensionFunction
 */
export function getOrNullByCollectionHolder<const T, >(collection: Nullable<CollectionHolder<T>>, index: number,): NullOr<T> {
    if (collection == null)
        return null
    if (Number.isNaN(index,))
        return null
    if (index == Number.NEGATIVE_INFINITY)
        return null
    if (index == Number.POSITIVE_INFINITY)
        return null
    if (collection.isEmpty)
        return null

    const size = collection.size
    if (index >= size)
        return null
    if (index >= 0)
        return collection.get(index,)

    const indexToRetrieve = index + size
    if (indexToRetrieve < 0)
        return null
    if (indexToRetrieve > size)
        return null
    return collection.get(indexToRetrieve,)

}

/**
 * Get the element at the specified index in the {@link collection}
 * or <b>null</b> if it is over the {@link size}
 *
 * @param collection The {@link Nullable nullable} {@link ReadonlyArray collection}
 * @param index      The index to retrieve a value
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/get-or-null.html Kotlin getOrNull(index)
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/element-at-or-null.html Kotlin elementAtOrNull(index)
 * @canReceiveNegativeValue
 * @extensionFunction
 */
export function getOrNullByArray<const T, >(collection: Nullable<readonly T[]>, index: number,): NullOr<T> {
    if (collection == null)
        return null
    if (Number.isNaN(index,))
        return null
    if (index == Number.NEGATIVE_INFINITY)
        return null
    if (index == Number.POSITIVE_INFINITY)
        return null

    const size = collection.length
    if (size == 0)
        return null
    if (index >= size)
        return null
    if (index >= 0)
        return collection[index] as T

    const indexToRetrieve = index + size
    if (indexToRetrieve < 0)
        return null
    return collection[indexToRetrieve] as T
}

//#endregion -------------------- Facade methods --------------------

File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/method/GetOrNull.java
package joookiwi.collection.java.method;

import joookiwi.collection.java.CollectionHolder;
import joookiwi.collection.java.MinimalistCollectionHolder;
import joookiwi.collection.java.annotation.CanReceiveNegativeValue;
import joookiwi.collection.java.annotation.ExtensionFunction;
import joookiwi.collection.java.exception.ImpossibleConstructionException;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Unmodifiable;

import static joookiwi.collection.java.CommonContracts.ALWAYS_FAIL_0;
import static joookiwi.collection.java.CommonContracts.IF_1ST_NULL_THEN_NULL_2;

@NotNullByDefault
public final class GetOrNull
        extends Utility {

    @Contract(ALWAYS_FAIL_0)
    private GetOrNull() { throw new ImpossibleConstructionException("The utility class “GetOrNull” cannot be constructed.", GetOrNull.class); }

    //#region -------------------- Facade methods --------------------

    /// Get the element at the specified `index` in the `collection`
    /// or `null` if it is over the [size][MinimalistCollectionHolder#size]
    ///
    /// @param collection The [nullable][Nullable] [collection][MinimalistCollectionHolder]
    /// @param index      The index to retrieve a value
    /// @param <T>        The `collection` type
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/get-or-null.html">Kotlin getOrNull(index)</a>
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/element-at-or-null.html">Kotlin elementAtOrNull(index)</a>
    @ExtensionFunction
    @CanReceiveNegativeValue
    @Contract(IF_1ST_NULL_THEN_NULL_2)
    public static <T extends @Nullable Object> @Nullable T getOrNull(final @Nullable MinimalistCollectionHolder<? extends @Nullable T> collection,
                                                                     final int index) {
        if (collection == null)
            return null;

        final var size = collection.size();
        if (size == 0)
            return null;
        if (index >= size)
            return null;
        if (index >= 0)
            return collection.get(index);

        final var indexToRetrieve = index + size;
        if (indexToRetrieve < 0)
            return null;
        return collection.get(indexToRetrieve);
    }

    /// Get the element at the specified `index` in the `collection`
    /// or `null` if it is over the [size][CollectionHolder#size]
    ///
    /// @param collection The [nullable][Nullable] [collection][CollectionHolder]
    /// @param index      The index to retrieve a value
    /// @param <T>        The `collection` type
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/get-or-null.html">Kotlin getOrNull(index)</a>
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/element-at-or-null.html">Kotlin elementAtOrNull(index)</a>
    @ExtensionFunction
    @Contract(IF_1ST_NULL_THEN_NULL_2)
    @CanReceiveNegativeValue
    public static <T extends @Nullable Object> @Nullable T getOrNull(final @Nullable CollectionHolder<? extends @Nullable T> collection,
                                                                     final int index) {
        if (collection == null)
            return null;
        if (collection.isEmpty())
            return null;

        final var size = collection.size();
        if (index >= size)
            return null;
        if (index >= 0)
            return collection.get(index);

        final var indexToRetrieve = index + size;
        if (indexToRetrieve < 0)
            return null;
        return collection.get(indexToRetrieve);
    }

    /// Get the element at the specified `index` in the `collection`
    /// or `null` if it is over the `size`
    ///
    /// @param collection The [nullable][Nullable] collection
    /// @param index      The index to retrieve a value
    /// @param <T>        The `collection` type
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/get-or-null.html">Kotlin getOrNull(index)</a>
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/element-at-or-null.html">Kotlin elementAtOrNull(index)</a>
    @ExtensionFunction
    @CanReceiveNegativeValue
    @Contract(IF_1ST_NULL_THEN_NULL_2)
    public static <T extends @Nullable Object> @Nullable T getOrNull(final T @Nullable @Unmodifiable [] collection,
                                                                     final int index) {
        if (collection == null)
            return null;

        final var size = collection.length;
        if (size == 0)
            return null;
        if (index >= size)
            return null;
        if (index >= 0)
            return collection[index];

        final var indexToRetrieve = index + size;
        if (indexToRetrieve < 0)
            return null;
        return collection[indexToRetrieve];
    }

    //#endregion -------------------- Facade methods --------------------

}
